kube-runtime/src/wait.rs

```diff
@@ -0,0 +1,95 @@
+use futures::TryStreamExt;
+use kube::{Api, Resource};
+use serde::de::DeserializeOwned;
+use snafu::{futures::TryStreamExt as _, Snafu};
+use std::fmt::Debug;
+
+use crate::watcher::{self, watch_object};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("failed to probe for whether the condition is fulfilled yet: {}", source))]
+    ProbeFailed {
+        #[snafu(backtrace)]
+        source: watcher::Error,
+    },
+}
+
+/// Watch an object, and Wait for some condition `cond` to return `true`.
+///
+/// `cond` is passed `Some` if the object is found, otherwise `None`.
+///
+/// # Errors
+///
+/// Fails if the type is not known to the Kubernetes API, or if the [`Api`] does not have
+/// permission to `watch` and `list` it.
+///
+/// Does *not* fail if the object is not found.
+pub async fn await_condition<K>(
+    api: Api<K>,
+    name: &str,
+    mut cond: impl FnMut(Option<&K>) -> bool,
+) -> Result<(), Error>
```

Comment:
Maybe, but it feels like we need to get a solid grasp of what abstraction level `Api` is supposed to work at first. At the very least, I'd promote `watcher` to `Api` before this (maybe even replacing the current `Api::watch`, or demoting that to `raw_watch` or similar?).

Reply:
> Maybe, but it feels like we need to get a solid grasp of what abstraction level Api is supposed to work at first.

Yeah, true. `Api` was meant to mirror the kubernetes api, so it would require a bit of care to make extensions for things that is more our abstractions. I think it would make sense to demote `Api::watch` to `::watch_raw` but the problem is that there are many replacements that are not straight analogues:

- `watcher` would kind of be a `::watch_collection`
- `watch_object` would kind of be `::watch`

But `watcher` is geared for reflectors, and requires a kube specific abstraction to understand. Maybe we should have a watch that doesn't rewrite events into reflector events, but is resilient against relists (at the cost of sometimes spamming the user with events)?

Similarly `watch_object` also rewrites a stream of `WatchEvent`s into a stream of `Option`s. This is nice for our internal consumers (`delete_and_finalize` plus `await_condition`), but requires understanding of those semantics externally. I feel that both our internal consumers could also be built against a resilient version of `Api::watch` (and it could do matching on WatchEvents internally, maybe with more re-usable methods on `WatchEvent`).

If we had this, then we could easily expose both of them on `Api` (as they are direct kubernetes api behaviour, which anyone coming from another language can pick up without much difficulty).

This would still leave the conveniences like `await_condition` and `delete_and_finalize` outside of it, unless we make them explicitly "not part of the kubernetes api" by adding them to a `WaitExt` trait (rather than a core `Wait` trait).

Reply:
> But watcher is geared for `reflectors`, and requires a kube specific abstraction to understand. Maybe we should have a watch that doesn't rewrite events into reflector events, but is resilient against relists (at the cost of sometimes spamming the user with events)?

I'm not sure I follow here, how'd you like to structure it instead then?

Reply:
I was thinking new watcher like thing that only deals with relists. I guess this is s bit much here. Maybe it's better if I do a follow up exploration pr on top of this?

Reply:
Sure.

Reply:
I am still playing around with this, have some changes locally, but stream was difficult and took time yesterday and won't get to do more until wednesday. I'll try to clarify again though for transparency (might be quicker to discuss it);

Essentially, was thinking we __could__ have a `watcher` like mechanism that does not emit `Event`s to the world, but rather straight `WatchEvent`s. For the initial list case, that means wrapping the output list elements in `WatchEvent::Added`, and in the re-list case, it means wrapping the output in `WatchEvent::Added`. The `step` machinery, I imagine, would be quite similar to before, but with an extra stream `flatten`ing going on in the `stream::unfold`.

It would mean it's not a great use-case for `reflector` though, and it would also mean duplication. But you would in effect have made a safer `Api::watch`. If a re-list happens, you kind of have to re-examine all objects anyway in a controller setting (because you might have missed modified events). I don't think the user gets a whole lot from being exposed to the explicit atomic-swap step (because they would need to scan everything object at that point).

Resource versions would be weird to track since you don't know when relists happen, but I think my point is that the user should not need to know about them.

Dealing with straight `WatchEvent`s can also give us a nice place to put common logic, e.g.:

```rust
impl<K: Resource> WatchEvent<K> {
    /// Check if a WatchEvent is one of the non-error types
    pub fn is_ok(&self) -> bool {
        !(matches!(&self, WatchEvent::Error(_)))
    }
    /// Check if a WatchEvent is a type that contains an object
    pub fn contains_object(&self) -> bool {
        !(matches!(&self, WatchEvent::Error(_) | WatchEvent::Bookmark(_)))
    }

    /// Extract the inner object of a WatchEvent (when it's not a bookmark or an error)
    pub fn inner(&self) -> Option<&K> {
        match &self {
            WatchEvent::Added(obj) | WatchEvent::Modified(obj) | WatchEvent::Deleted(obj) => {
                Some(obj)
            },
            _ => None
        }
    }
}
```

which should lend itself to common filtering.

Reply:
That sounds an awful lot like what you'd get by piping the watcher through [`try_flatten_applied`](https://docs.rs/kube-runtime/0.60.0/kube_runtime/utils/fn.try_flatten_applied.html) or [`try_flatten_touched`](https://docs.rs/kube-runtime/0.60.0/kube_runtime/utils/fn.try_flatten_touched.html)? (Although those names are evidently awful.. :P)

Anyway.. if we're not exposing the resync `resourceVersion`s then we probably shouldn't expose `Bookmark` events either (since trying to keep track of `resourceVersion` is kind of a lost cause at that point).

Reply:
It is possible to build something resembling the current `Api::watch` API by wrapping the current implementation, but there are some impedence mismatches that make me question whether returning `WatchEvent` is really the right choice: https://gist.github.com/teozkr/6d159e2481fd5db5883e8eb8db7b97fa

- (Design choice) `Event::Applied` has no idea whether it comes from `WatchEvent::Added` or `WatchEvent::Modified`
- (Inherent) `Event::Restarted` has no idea whether the items it lists are `Added` or `Modified`
- (Inherent) `Event::Restarted` has no idea what the deleted items are, so we can't generate `WatchEvent::Deleted` events
- (Inherent..ish) `watcher::Error` covers a lot more ground than `WatchEvent::Error`, so we still need to return a `Result`

Reply:
Ok, I have unfortunately not had a lot of time to rethink this, but it's ultimately not clear if an alternative interface here really would be better.  Also it's not right to hold this up when it's definitely good enough, so resolving this for now. Might look into a follow-up issue later down the line. Will make some minor comments on naming.